File: lib_book/src/main/java/com/ebook/basebook/mvp/view/adapter/ImportBookAdapter.java

package com.ebook.basebook.mvp.view.adapter;

import android.annotation.SuppressLint;
import android.os.Environment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;


import com.ebook.basebook.R;
import com.ebook.basebook.view.checkbox.SmoothCheckBox;

import java.io.File;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class ImportBookAdapter extends RecyclerView.Adapter<ImportBookAdapter.Viewholder> {
    private final List<File> fileList;
    private final List<File> selectFileList;

    public interface OnCheckBookListener {
        void checkBook(int count);
    }

    private final OnCheckBookListener checkBookListener;

    public ImportBookAdapter(@NonNull OnCheckBookListener checkBookListener) {
        fileList = new ArrayList<>();
        selectFileList = new ArrayList<>();
        this.checkBookListener = checkBookListener;
    }

    @NonNull
    @Override
    public Viewholder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new Viewholder(LayoutInflater.from(parent.getContext()).inflate(R.layout.view_adapter_importbook, parent, false));
    }

    @Override
    public void onBindViewHolder(final Viewholder holder, final int position) {
        holder.tvNmae.setText(fileList.get(position).getName());
        holder.tvSize.setText(convertByte(fileList.get(position).length()));
        holder.tvLoc.setText(fileList.get(position).getAbsolutePath().replace(Environment.getExternalStorageDirectory().getAbsolutePath(), "存储空间"));

        holder.scbSelect.setOnCheckedChangeListener((checkBox, isChecked) -> {
            if (isChecked) {
                selectFileList.add(fileList.get(position));
            } else {
                selectFileList.remove(fileList.get(position));
            }
            checkBookListener.checkBook(selectFileList.size());
        });
        if (canCheck) {
            holder.scbSelect.setVisibility(View.VISIBLE);
            holder.llContent.setOnClickListener(v -> holder.scbSelect.setChecked(!holder.scbSelect.isChecked(), true));
        } else {
            holder.scbSelect.setVisibility(View.INVISIBLE);
            holder.llContent.setOnClickListener(null);
        }
    }

    public void addData(File newItem) {
        int position = fileList.size();
        fileList.add(newItem);
        notifyItemInserted(position);
        notifyItemRangeChanged(position, 1);
    }

    private Boolean canCheck = false;

    @SuppressLint("NotifyDataSetChanged")
    public void setCanCheck(Boolean canCheck) {
        this.canCheck = canCheck;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return fileList.size();
    }

    static class Viewholder extends RecyclerView.ViewHolder {
        LinearLayout llContent;
        TextView tvNmae;
        TextView tvSize;
        TextView tvLoc;
        SmoothCheckBox scbSelect;

        public Viewholder(View itemView) {
            super(itemView);
            llContent = itemView.findViewById(R.id.ll_content);
            tvNmae = itemView.findViewById(R.id.tv_name);
            tvSize = itemView.findViewById(R.id.tv_size);
            scbSelect = itemView.findViewById(R.id.scb_select);
            tvLoc = itemView.findViewById(R.id.tv_loc);
        }
    }

    public static String convertByte(long size) {
        DecimalFormat df = new DecimalFormat("###.#");
        float f;
        if (size < 1024) {
            f = size;
            return (df.format(Float.valueOf(f).doubleValue()) + "B");
        } else if (size < 1024 * 1024) {
            f = (float) size / (float) 1024;
            return (df.format(Float.valueOf(f).doubleValue()) + "KB");
        } else if (size < 1024 * 1024 * 1024) {
            f = (float) size / (float) (1024 * 1024);
            return (df.format(Float.valueOf(f).doubleValue()) + "MB");
        } else {
            f = (float) size / (float) (1024 * 1024 * 1024);
            return (df.format(Float.valueOf(f).doubleValue()) + "GB");
        }
    }

    public List<File> getSelectFileList() {
        return selectFileList;
    }

    public List<File> getFileList(){return fileList;}
}

File: module_book/src/main/java/com/ebook/book/adapter/ImportBookAdapter.kt
package com.ebook.book.adapter

import android.annotation.SuppressLint
import android.content.Context
import android.os.Environment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.ebook.common.callback.FileDifferCallback
import com.ebook.common.databinding.ViewAdapterImportbookBinding
import com.ebook.common.view.checkbox.SmoothCheckBox
import com.xrn1997.common.adapter.BaseBindAdapter
import java.io.File
import java.text.DecimalFormat

class ImportBookAdapter(context: Context) :
    BaseBindAdapter<File, ViewAdapterImportbookBinding>(context, FileDifferCallback()) {
    private val selectFileList: MutableList<File> =
        ArrayList()
    private var canCheck = false

    private var mCheckBookListener: ((count: Int) -> Unit)? = null

    @SuppressLint("NotifyDataSetChanged")
    fun setCanCheck(canCheck: Boolean) {
        this.canCheck = canCheck
        notifyDataSetChanged()
    }

    override fun onBindItem(binding: ViewAdapterImportbookBinding, item: File, position: Int) {
        binding.tvName.text = item.name
        binding.tvSize.text =
            convertByte(item.length())
        binding.tvLoc.text = item.absolutePath.replace(
            Environment.getExternalStorageDirectory().absolutePath,
            "存储空间"
        )

        binding.scbSelect.setOnCheckedChangeListener { _: SmoothCheckBox?, isChecked: Boolean ->
            if (isChecked) {
                selectFileList.add(item)
            } else {
                selectFileList.remove(item)
            }
            mCheckBookListener?.invoke(selectFileList.size)
        }
        if (canCheck) {
            binding.scbSelect.visibility = View.VISIBLE
            binding.llContent.setOnClickListener {
                binding.scbSelect.setChecked(
                    !binding.scbSelect.isChecked,
                    true
                )
            }
        } else {
            binding.scbSelect.visibility = View.INVISIBLE
            binding.llContent.setOnClickListener(null)
        }

    }

    override fun onBindViewBinding(
        inflater: LayoutInflater,
        parent: ViewGroup,
        attachToParent: Boolean,
        viewType: Int
    ): ViewAdapterImportbookBinding {
        return ViewAdapterImportbookBinding.inflate(inflater, parent, attachToParent)
    }

    fun getSelectFileList(): List<File> {
        return selectFileList
    }

    fun setCheckBookListener(checkBookListener: (count: Int) -> Unit) {
        mCheckBookListener = checkBookListener
    }

    companion object {
        fun convertByte(size: Long): String {
            val df = DecimalFormat("###.#")
            val f: Float
            if (size < 1024) {
                f = size.toFloat()
                return (df.format(f.toDouble()) + "B")
            } else if (size < 1024 * 1024) {
                f = size.toFloat() / 1024f
                return (df.format(f.toDouble()) + "KB")
            } else if (size < 1024 * 1024 * 1024) {
                f = size.toFloat() / (1024 * 1024).toFloat()
                return (df.format(f.toDouble()) + "MB")
            } else {
                f = size.toFloat() / (1024 * 1024 * 1024).toFloat()
                return (df.format(f.toDouble()) + "GB")
            }
        }
    }
}
